src/freertos_drivers/st/Stm32Gpio.hxx

```diff
@@ -192,25 +191,252 @@ template <class Defs, uint32_t PULL_MODE> struct GpioInputPin : public Defs
 ///
 /// Do not use this class directly. Use @ref GPIO_PIN instead.
 template <class Defs>
 struct GpioInputNP : public GpioInputPin<Defs, GPIO_NOPULL>
+{ };
+
+#include "utils/OptionalArgs.hxx"
+
+/// Configuration options for an STM32 GPIO pin.
+struct Stm32GpioOptionDefs
 {
+    /// Mode: GPIO_MODE_xx value, including GPIO_MODE_INPUT,
+    /// GPIO_MODE_OUTPUT_PP, GPIO_MODE_OUTPUT_OD, GPIO_MODE_AF_PP,
+    /// GPIO_MODE_AF_OD.
+    DECLARE_OPTIONALARG(GpioMode, gpio_mode, uint32_t, 0, GPIO_MODE_INPUT);
+    /// Pullup/Pulldown configuration. One of GPIO_NOPULL, GPIO_PULLUP,
+    /// GPIO_PULLDOWN.
+    DECLARE_OPTIONALARG(Pull, pull, uint32_t, 2, GPIO_NOPULL);
+    /// Speed configuration. One of GPIO_SPEED_FREQ_LOW,
+    /// GPIO_SPEED_FREQ_MEDIUM, GPIO_SPEED_FREQ_HIGH, GPIO_SPEED_FREQ_VERYHIGH.
+    DECLARE_OPTIONALARG(Speed, speed, uint32_t, 3, GPIO_SPEED_FREQ_LOW);
+    /// Used for selecting alternate function. Many values are possible, for
+    /// example GPIO_AF0_UART3 or GPIO_AF3_FDCAN1. Typically AF0 to AF7, or on
+    /// some chips until AF15, and the peripheral name can be a large
+    /// variety. Take the values from the datasheet or the specific chip's
+    /// header file.
+    DECLARE_OPTIONALARG(AfMode, afmode, uint32_t, 4, 0xffffffff);
+
+    /// Sets the GPIO port base, like GPIOA_BASE, GPIOB_BASE, etc. Required.
+    DECLARE_OPTIONALARG(PeriphBase, periph_base, uint32_t, 10, 0xffffffff);
+    /// Sets the GPIO pin number, like GPIO_PIN_0, GPIO_PIN_1, etc. Required.
+    DECLARE_OPTIONALARG(Pin, pin, uint16_t, 11, 0xffff);
+    /// Sets the GPIO pin number, 0..15. Required.
+    DECLARE_OPTIONALARG(PinNum, pin_num, uint8_t, 11, 0xff);
+
+    using Base = OptionalArg<Stm32GpioOptionDefs, GpioMode, Pull, Speed,
+        AfMode, PeriphBase, Pin, PinNum>;
 };
 
+class Stm32GpioOptions : public Stm32GpioOptionDefs::Base
+{
+public:
+    INHERIT_CONSTEXPR_CONSTRUCTOR(Stm32GpioOptions, Stm32GpioOptionDefs::Base);
+
+    DEFINE_OPTIONALARG(GpioMode, gpio_mode, uint32_t);
+    DEFINE_OPTIONALARG(Pull, pull, uint32_t);
+    DEFINE_OPTIONALARG(Speed, speed, uint32_t);
+    DEFINE_OPTIONALARG(AfMode, afmode, uint32_t);
+
+    DEFINE_OPTIONALARG(PeriphBase, periph_base, uint32_t);
+    DEFINE_OPTIONALARG(Pin, pin, uint16_t);
+    DEFINE_OPTIONALARG(PinNum, pin_num, uint8_t);
+
+    constexpr bool is_af() const
+    {
+        return (gpio_mode() == GPIO_MODE_AF_PP) ||
+            (gpio_mode() == GPIO_MODE_AF_OD);
+    }
+
+    constexpr bool is_od() const
+    {
+        return (gpio_mode() == GPIO_MODE_AF_OD) ||
+            (gpio_mode() == GPIO_MODE_OUTPUT_OD);
+    }
+
+    constexpr void __attribute__((always_inline))
+    fill_options(GPIO_InitTypeDef *gpio_init)
+    {
+        return fill_options(gpio_init, *this);
+    }
+
+    template <typename... Args>
+    static constexpr void __attribute__((always_inline))
+    fill_options(GPIO_InitTypeDef *gpio_init, Args... args)
+    {
+#if 0
```

Comment:
Is this dead code? Should there be a block comment?

Reply:
I added a big comment on why this code is deactivated. I'd like to have it in there but it doesn't compile and I don't know how to make it compile.